### DISCUSSION TITLE
`shift-heading-level-by` vs `reference-section-title`

### DISCUSSION THREAD

**Discussion by p3palazzo** [jgm/pandoc#9664] [Q&A] [answered]
It seems the `shift-heading-level-by` is not applied to the reference section title when the latter has been generated from metadata. Is this the expected behavior? I think it would make sense if the references title were subjected to the same heading level shift as the other headings.

**Comment by jgm** [comment, marked as answer]
That's right.  `--shift-heading-level-by` is implemented as a "transform" in <https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/App.hs#L296-L309>.
(The other "transforms" are handlers for `east_asian_line_breaks` and ipynb output.)  Currently the transforms are applied right before filters (and thus before citeproc).  I think it probably makes sense to move the transforms *after* the filters.  @tarleb can you think of anything this would break?

**Reply by bp0qd** [reply]
This may break the cross-reference numbers generated by pandoc-crossref. Specifically, when Pandoc uses both the "--shift-heading-level-by" and "--filter pandoc-crossref" options, if "--shift-heading-level-by" is executed after pandoc-crossref and the input file contains only level 2 headings and below, pandoc-crossref will generate incorrect cross-reference numbers, because it cannot see the level 1 headings.

For example, the following markdown will generate “Figure 0.1” instead of the expected “Figure 1.1”

``` markdown
---
chapters: true
---
## A Second-Level Heading {sec:SecondLevelHeading}

![Image](img.png){#fig:test}

See [@fig:test].
```

when I run the following command

``` bash
pandoc --top-level-division=chapter --shift-heading-level-by=-1 --filter=pandoc-crossref --to=html test.md --output=test.html
```

The output is

```html
<h1 id="sec:SecondLevelHeading">A Second-Level Heading</h1>
<figure id="fig:test">
<img src="img.png" alt="Figure 0.1: Image" />
<figcaption aria-hidden="true">Figure 0.1: Image</figcaption>
</figure>
<p>See fig. 0.1.</p>

```

**Reply by jgm** [reply]
Filters are applied ~~after~~ before transforms. [EDIT: you were right, they are applied before]

**Reply by jgm** [reply]
I think it would make sense to change this.  The "transforms" do things that the filters should have access to.  There's a chance of breaking something, though, if we do change it.

**Reply by jgm** [reply]
Hm, found this in git, I forgot the context of this discussion!
```
commit 3d902349d1f8fd9a54539f8e16772c781dbd1063
Author: John MacFarlane <jgm@berkeley.edu>
Date:   Wed Apr 17 22:48:56 2024 -0700

    T.P.App: Move "transforms" after filters.

    This will mean that `--shift-heading-level-by` affects a
    heading added by `reference-section-title`. See #9664.
```

**Reply by bp0qd** [reply]
I've already implemented a Lua filter as a workaround to handle the header level shifting before pandoc-crossref runs, and it’s working perfectly.

I completely understand the concern about potential side effects if the internal execution order is changed. In that case, would you consider adding a note to the documentation instead? It could help other users avoid this pitfall.

**Reply by jgm** [reply]
In light of this, I'm tempted to revert this change (which was made in pandoc 3.2).  The issue of reference-section-title seems less serious than the pandoc-crossref interaction; it is easy enough to just specify a refs section manually.

But a workaround would be to use two runs of pandoc (the first producing and the second consuming json), running the filters on the second and `--shift-heading-level-by` on the first.

**Reply by jgm** [reply]
Yes, a note to the docs would be considered. Where would it go, on `--shift-heading-level-by`?

**Reply by cagix** [reply]
that would break my workflows. running pandoc twice would be really annoying :(

**Reply by bp0qd** [reply]
> In light of this, I'm tempted to revert this change (which was made in pandoc 3.2). The issue of reference-section-title seems less serious than the pandoc-crossref interaction; it is easy enough to just specify a refs section manually.
> 
> But a workaround would be to use two runs of pandoc (the first producing and the second consuming json), running the filters on the second and `--shift-heading-level-by` on the first.

Yes, this workaround would work, but I’ve decided against it. Modifying our CMake build scripts is more challenging than simply adding a Lua filter to the existing commands. Now the filter is working properly, everything is fine on my end.

If you decide to revert the change, I will remove the filter and re-enable `--shift-heading-level-by` option.


> Yes, a note to the docs would be considered. Where would it go, on `--shift-heading-level-by`?

Yes, adding the note on `--shift-heading-level-by` is perfect.

**Reply by jgm** [reply]
>  running pandoc twice would be really annoying

It's still a one-liner with a pipe.

**Reply by cagix** [reply]
that would be fine for simple conversion, e.g. one document. but consider converting some hundreds and somewhat larger documents, this will become time consuming. also the makefiles get more complicated and more difficult to read.

**Comment by tarleb** [comment]
There might be some filters that would break, but that's about it. No objections to the change from my side.

**Reply by cagix** [reply]
such a change would easily annoy people who rely on the title metadata becoming the new h1 heading visible to the filter when using "--shift-heading-level-by=1". 

i use this in my filters quite often. of course, you could always search yourself for a title metadata, but that would be a rather clumsy workaround imho.

**Reply by jgm** [reply]
To be clear, the change being discussed was already made, in 2024.

**Comment by jgm** [comment]
This discussion is a bit confusing, so I'm going to summarize here.

In 2024, we made the change discussed in the original post above:  we moved transforms *after* filters (previously they'd been applied before filters).  The motivation was to allow `--shift-heading-level-by` to affect a title added by `citeproc` from the `reference-section-title` metadata field.

In 2026, it was pointed out that this change had a bad side effect when pandoc is used with the popular `pandoc-crossref` filter.  Since the filter now sees the section levels before they are transformed by `--shift-heading-level-by`, they get inappropriate numbering.

Both of these are legitimate issues, and I'm not really sure how to handle them.  Both can be worked around.

I think that conceptually, it makes a bit more sense to do transforms before filters, as they were originally conceived.  The transforms are built-in alterations to parsing, which one might think should be in place when filters get to look at the AST.

A fancier model would be this:  we could treat the `--shift-heading-level-by` as a built-in filter, as we currently treat `--citeproc`, and this would allow the user to order it arbitrarily with respect to other filters. I haven't looked into this option but it is worth exploring, I think.

If we don't implement that, or revert the original change, then we should probably add a note to the manual for `--shift-heading-level-by` to indicate that the transformation takes place after filters are applied.

**Reply by cagix** [reply]
> I think that conceptually, it makes a bit more sense to do transforms before filters, as they were originally conceived. The transforms are built-in alterations to parsing, which one might think should be in place when filters get to look at the AST.

Couldn't the parser just do the parsing, with the transformations kept separate from the parser wherever possible? I've run into some odd situations in the past because certain processing steps were already taking place during the parsing stage. 

> 
> A fancier model would be this: we could treat the `--shift-heading-level-by` as a built-in filter, as we currently treat `--citeproc`, and this would allow the user to order it arbitrarily with respect to other filters. I haven't looked into this option but it is worth exploring, I think.

I think it would be more useful to treat all(!) transformations in the same way as filters (similar to cite-proc). If users simply use the CMD option ‘--shift-heading-level-by’, these filters could be applied in the same order as they are now, so no workflows would be disrupted. If users specify the option as a filter (e.g. via the filter section in default files), they could choose the execution order freely and thus deliberately achieve the desired behaviour.

**Reply by jgm** [reply]
Besides `--shift-heading-level-by`, there are only two things that are currently treated as transforms: `east_asian_line_breaks` (which is triggered by a reader extension and should presumably go before the filters -- it's just implemented as a transform for convenience, so I don't have to modify every reader), and ipynb output options (which really just affects conversions from ipynb, but must be here rather than the reader because we need to know the output format).  I think we could leave those as they are and just change the heading level shift to a kind of built-in filter, like citeproc.

**Reply by bp0qd** [reply]
> Besides `--shift-heading-level-by`, there are only two things that are currently treated as transforms: `east_asian_line_breaks` (which is triggered by a reader extension and should presumably go before the filters -- it's just implemented as a transform for convenience, so I don't have to modify every reader), and ipynb output options (which really just affects conversions from ipynb, but must be here rather than the reader because we need to know the output format). I think we could leave those as they are and just change the heading level shift to a kind of built-in filter, like citeproc.

I see what you mean about the other two transforms, but since you’re already considering making the header shift a built-in filter, it might eventually be worth bringing `east_asian_line_breaks` and the ipynb output options into that same framework too.

From a workflow consistency perspective, having all AST modifications handled as filters would make the 'Reader → Filter → Writer' pipeline very clean and predictable. It’s just a thought for the long-term architecture, but I think it would make the whole system even more robust.

Of course, I realize that might be a bigger task, so it's entirely up to you. Thank you for making such great software!

**Comment by massifrg** [comment]
In the command line, filters are run in the order in which they are specified.

If only the `--shift-heading-level-by` -- as other transforms that can be seen as internal filters -- could be run in the order in which they are specified on the command line, the user could then choose to run them before or after (or even between) other filters.

**Reply by massifrg** [reply]
Something like [this](https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/Filter.hs#L40-L44):
```haskell
-- | Type of filter and path to filter file.
data Filter = LuaFilter FilePath
            | JSONFilter FilePath
            | CiteprocFilter -- built-in citeproc
            | TransformFilter Text Transform -- a transform with a name
            deriving (Show, Generic, Eq)
```
and [then](https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/Filter.hs#L76-L101):
```haskell
applyFilters scrngin fenv filters args d = do
  expandedFilters <- mapM expandFilterPath filters
  foldM applyFilter d expandedFilters
 where
  applyFilter doc (JSONFilter f) =
    withMessages f $ JSONFilter.apply fenv args f doc
  applyFilter doc (LuaFilter f)  =
    withMessages f $ engineApplyFilter scrngin fenv args f doc
  applyFilter doc CiteprocFilter =
    withMessages "citeproc" $ processCitations doc
  applyFilter doc (TransformFilter name transform) =
    withMessages ("transform " <> name) $ applyTransform doc  -- it must be fixed, this is just to illustrate the idea
```
because:
```haskell
applyFilter :: Pandoc -> Filter -> m Pandoc
Transform :: Pandoc -> Pandoc
```
Monads aside, both Filter and Transform are essentially `Pandoc -> Pandoc` functions.
Leading transforms back to filters, a transform can be applied in any order among filters.
Something should be fixed in the CLI specs, but this would allow shifting headers levels both before and after other filters.

**Reply by jgm** [reply]
Yes, this is just the sort of thing I was suggesting in the comment above.
